### PR TITLE
Fix #39: Add nonlinear one-step trainer phase

### LINKED ISSUE #39: Add nonlinear one-step trainer phase
## Summary
Add a new one-step trainer for nonlinear models alongside the existing linear trainer.

## Problem
The current one-step-style training phase appears to be specialized for linear models because it forms one-step targets and then solves a linear system. That does not extend to nonlinear models.

We need an equivalent training path that keeps the same target construction idea:
- For discrete-time models, predict the next state.
- For continuous-time models, predict the rate of change.

But instead of solving a linear system, the nonlinear variant should fit the model with SGD-style optimization.

## Expected outcome
Introduce a nonlinear one-step trainer phase that:
- Mirrors the existing one-step target-generation logic where appropriate.
- Trains nonlinear models with gradient-based optimization rather than linear solves.
- Can be used as a standalone training phase.
- Can also participate in stacked phase workflows, such as warm-starting a NODE trainer.

## Acceptance criteria
- There is a new training phase for nonlinear one-step training, implemented in parallel to the existing linear trainer rather than overloading it with incompatible logic.
- The phase supports both target modes already implied by the current trainer design:
  - discrete-time next-state targets
  - continuous-time rate targets
- The training path uses an SGD-type optimizer instead of forming and solving a linear system.
- The phase can be configured and executed as a standalone trainer.
- The phase can be composed in stacked training workflows, with at least one covered warm-start use case for a downstream nonlinear trainer (for example, NODE), if that composition model already exists in the codebase.
- Tests cover the new phase at the appropriate layer for this repo.
- If this changes training-phase boundaries or recommended feature placement, update the relevant architecture/placement docs in the same change.

## Notes
- Keep this separate from the linear trainer; the goal is a parallel nonlinear path, not a partial generalization that preserves linear-system assumptions.
- The exact API, naming, and optimizer scope are still open and should follow the existing training-phase structure in the repo.
- It is not yet specified whether the new trainer should share a base abstraction with the linear trainer or simply match its external behavior.

### PR DESCRIPTION
**Pull request by smanist**
Closes #39

## Summary
This PR was generated by the local AI issue worker.

## Verification
PASS make check
exit code: 0
duration: 4.37s
stdout:
ruff check .
All checks passed!
ruff format . --check
264 files already formatted
pyright --pythonpath "####/miniconda3/bin/python"
0 errors, 0 warnings, 0 informations

stderr:

PASS pytest -m "not slow and not extra_slow"
exit code: 0
duration: 91.48s
stdout:
ript.py:1488: DeprecationWarning: `torch.jit.script` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
tests/test_assert_krr.py::test_krr
  ####/miniconda3/lib/python3.12/site-packages/torch/_tensor.py:1120: DeprecationWarning: __array_wrap__ must accept context and return_scalar arguments (positionally) in the future. (Deprecated NumPy 2.0)
    return self.reciprocal() * other

tests/test_workflow_sa_lti.py::test_sa[1]
  ####/Repos/dymad-dev/.ai-worktrees/issue-39/src/dymad/training/ls_update.py:368: UserWarning: Casting complex values to real discards the imaginary part (Triggered internally at ####/work/pytorch/pytorch/pytorch/aten/src/ATen/native/Copy.cpp:308.)
    Wt = torch.tensor(W, dtype=dtype, device=device)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
========== 394 passed, 71 deselected, 29 warnings in 87.76s (0:01:27) ==========

stderr:

## Changed files
- src/dymad/agent/exec/workflow.py
- src/dymad/agent/registry/training_schema.py
- src/dymad/training/__init__.py
- src/dymad/training/ls_update.py
- src/dymad/training/phase_runtime.py
- src/dymad/training/phases.py
- src/dymad/training/trainer.py
- tests/test_agent_mcp_user_tools.py
- tests/test_agent_registry.py
- tests/test_agent_training_compiler.py
- tests/test_contract_public_reexport_surfaces.py
- tests/test_contract_training_phase_runtime.py
- tests/test_workflow_kp.py

## Agent notes
See diff for implementation details.

### CODE REVIEW

**Comment by smanist** [comment]
Done elsewhere
